networked-hand-controls, with gestures; also, many examples updates

kylebakerio opened the pull request:
Based on the A-Frame built-in `hand-controls` component. Heavily refactored for readability, performance, and functionality. Will likely make an upstream pull request to A-Frame based on these updates in the near future.

Of course, added the ability to integrate with NAF as well. Gesture updates are stored as a schema prop, and so are easy to propagate via NAF. tracked controller demo has been updated to use this component, as well. dist file has been compiled with the new component, which is necessary to be able to use the file without duplicating it.

kylebakerio commented:
To test locally, I recommend using the webxr chrome extension. Also, though, feel free to ping me to try it out.

kylebakerio commented:
"eslint": "^6.6.0", in package.json, I presume that's what we're using? [eslint added support](https://stackoverflow.com/questions/61628947/eslint-optional-chaining-error-with-vscode) for conditional chaning in 7.5

vincentfretin commented on src/components/networked-hand-controls.js:
```diff
@@ -0,0 +1,356 @@
+      // ^while this would be more ideal, NAF doesn't populate the owner until later
+      // and we need to know this at init(), and we can't supply schema options
+      // when instantiating a template instance--so, we require adding a class for now.
+      this.el.classList.forEach(aClass => {
```
`classList` has a `contains` api that may simplify this code to
`this.local = this.el.classList.contains("local-naf-hand")`
https://developer.mozilla.org/en-US/docs/Web/API/Element/classList

kylebakerio replied:
I tried to use that, but it wasn't available here. I think A-Frame doesn't pass it along? I'll try again and see if I had some random weirdness going on, but that was my logic here.

kylebakerio replied:
Seems to work now, dunno what was happening at the moment I was coding it. Updated.

vincentfretin commented on src/components/networked-hand-controls.js:
```diff
@@ -0,0 +1,356 @@
+    },
+    
+    update(oldData) {   
+      console.log(oldData, this.data)
```
console.log to be removed

kylebakerio replied:
thanks, will do

vincentfretin commented on src/components/networked-hand-controls.js:
```diff
@@ -0,0 +1,356 @@
+      this.loader.load(this.data.handModelURL || handmodelUrl, 
+                       gltf => {
+        const newMesh = gltf.scene.children[0];
+        // const mesh = gltf.scene.children[0];
```
commented line to be removed

vincentfretin commented on src/components/networked-hand-controls.js:
```diff
@@ -0,0 +1,356 @@
+        
+        this.el.addEventListener('controllerconnected', () => {
+          this.el.setAttribute('networked-hand-controls', 'visible', true);
+          this.el.object3D.visible = true;
```
This line may be unnecessary if your update() method is already handling this.

kylebakerio replied:
Great point! Original component had this, I later added the update handler for NAF, didn't catch the overlap.

vincentfretin reviewed (changes requested):
That's really cool! I tested locally with the PC and the Quest by defining the certificate for https in `server/easyrtc-server.js`. I couldn't make the WebXR Chrome extension work.

Last time I resaved all the examples from vscode with the prettier config in the repo. Can you please open the html file in vscode and resave?

For conditional chaining, you can try updating eslint, be sure there is no error when you run `npm run lint`. Otherwise remove your usage of conditional chaining and replace it by a if.

Please add some documentation about the component in the README and address my comments and it will be good to merge. Thanks for this contribution!

vincentfretin commented:
![Capture d’écran de 2022-08-08 11-51-44](https://user-images.githubusercontent.com/112249/183391140-86d7186f-e924-4b5c-a93d-21a5d3ef0b20.png)
![Capture d’écran de 2022-08-08 11-51-51](https://user-images.githubusercontent.com/112249/183391144-b0108a24-3b24-44ff-b20e-e7c7cb84fd71.png)

kylebakerio commented:
Sorry for the scope creep... As I went to write out how to implement the new component, I was dissatisfied and realized the interface needed to be improved. In the process, I went ahead and cleaned up some other stuff in the examples section as well.

- added requested readme updates, with documentation and guide on adding tracked controllers
- updated readme list of examples to include video, and to separate updated/non-updated examples, and updated audio glitch link
- updated basic example by pulling relevant sections from the previous version of tracked-controlers.html
- simplified color tracking in basic example
- vastly simplified tracked-controllers.html to be a minimal example that only demonstrates how to add tracked controllers--so, moved all basic comments to the basic example, and removed the nametag syncing functionality
- updated networked-hand-controls to respect color specified in schema, which has been broken in hand-controls component
- updated networked-hand-controls to have a _much_ simpler user interface--it now updates the NAF template cache in a different way that avoids awkward workarounds within the HTML
- added separate nametag example, and added random-color-update button;

vincentfretin commented:
Most of the changes are fine. I added comments directly on commit https://github.com/networked-aframe/networked-aframe/commit/28dc8995321c278a292cf4acb6604095fe65f8a4 but it doesn't show up here in the PR.

vincentfretin commented:
You should remove the player-info component in basic.html, it's not used.

vincentfretin commented:
Please add the new nametag example in index.html

vincentfretin reviewed (approved):
I think it's ready to be merged now. Or do you still need to work on it?

kylebakerio commented:
I agree, ready when you are.

kylebakerio commented:
remember to go update the naf-examples glitch! cheers, thanks for reviewing all that.
